### ISSUE TITLE
Make -every not emit marks (again)

### ISSUE THREAD

**Issue by dmehra** [juttle/juttle#594]
This is in question / maybe-bug / gotta figure this out category. When we recently made a change to pass the batch marks from an upstream `reduce -every` to downstream procs, some cases were not considered, that now showed up in practice, such as:

- if upstream reduce has `-every` but downsteam of it there is a reduce without `-every` that's meant to aggregate all the data, it now aggregates on the same `-every` interval as upstream, and requires introduction of `unbatch` into the program to get back to the intended behavior. I had to do this in [postgres-diskstats/throughput.juttle](https://github.com/juttle/juttle-engine/pull/87/files#diff-9a9a1b96627402f2d5161a5726c3212c) example. (Why the two-level reduce: because there's a lot of raw data in the backend, so I'm doing a `reduce -every :hour:` on read that gets optimized with the read, and then i fork the program to do something with hourly statistics, and also aggregate across.)

- similarly if upstream reduce has `-every` but downstream of it you want to sort across all data, `reduce -every :day: count() | sort count` to get top daily counts, that no longer works and needs to be `reduce -every :day: count() | unbatch | sort count`. 

Those are common use cases, and we have made them more complex for the user, who now has to be aware of `unbatch` even though they never used `batch`.

The original idea was to simplify the case where the user continues doing calculations for the same batch interval, such as `reduce -every :hour: ... | head 1` and means to get the head of every hour. That is still a case we want to support, but not at the expense of complicating the also-common cases above. 

Options:

(1) Drop the support for `-every` and instead use `batch` (and explicit `unbatch` where needed). See @go-oleg.

(2) Introduce parameter `-all true` (unless there's a better name) that acts as `unbatch` but applies to reduce, head, sort and anything that supports `-every`, so in the sort example above, the user would write `reduce -every :day: count() | sort -all true count`. I wish we didn't have to give it 'true' and could just say `sort -all count`. This was @dmehra's vote. 

(3) Similar to (2) but use `-every :forever:` which should be already supported logically (the parser barfs today). Suggested by @demmer but perhaps not seriously.

(4) Rename `unbatch` to something friendlier. @go-oleg said `unevery`. Maybe even friendlier. Don't know what.

(5) Go back to not passing the batch marks downstream. That will break some use case we wanted to make easy, but we (@dmehra, @go-oleg) failed to remember exactly what that was.

For your consideration, @demmer @dmajda and general interest @juttle/developers .

**Comment by rlgomes** [comment]
Seems like our default -every should be `:all time:` so that when you do `reduce count()` this is in fact a `reduce -every :alltime: count()` and then the behavior would be the expected one:

```
rlgomes@x230> ./bin/juttle -e "emit -from :0: -limit 100 -every :1s: | reduce -every :5s: count() | reduce -every :100y: count()"
┌────────────────────────────────────┬──────────┐
│ time                               │ count    │
├────────────────────────────────────┼──────────┤
│ 2070-01-01T00:00:00.000Z           │ 20       │
└────────────────────────────────────┴──────────┘
```

being that we don't have `:alltime:` I set the every to 100 years.

**Comment by demmer** [comment]
For the record re (3) it's not the parser that complains about `-every :forever:` but the runtime. And no it was not a very serious suggestion.

**Comment by bkutil** [comment]
Here's the original bug report and discussion: https://github.com/juttle/juttle/issues/52

**Comment by dmajda** [comment]
Two notes while I’m thinking about this:

> - similarly if upstream reduce has `-every` but downstream of it you want to sort across all data, `reduce -every :day: count() | sort count` to get top daily counts, that no longer works and needs to be `reduce -every :day: count() | unbatch | sort count`.
 
@dmehra Interestingly, this is almost exactly the example @mccanne used few months ago to convince me that `reduce -every` should emit marks. His intuition was that `sort` should operate on batches in this case, not on the whole stream (to him, the need to add additional `batch` or an `-every` option to `sort` was the bad thing).

> Seems like our default -every should be `:all time:` so that when you do `reduce count()` this is in fact a `reduce -every :alltime: count()` and then the behavior would be the expected one:

@rlgomes With just `reduce` with would be fine but we want to add `-every` also to other batch-aware procs such as `head` or `sort` (see #53). Since we want `-every` to behave the same everywhere, they’d also have to have “all time” as a default value, meaning they would effectively ignore batches. The result would be almost equivalent with not having batches in the stream at all.

**Comment by dmehra** [comment]
My interpretation of https://github.com/juttle/juttle/issues/52 was that it would fix the sinks' understanding of batch marks, resolving the problem we had with `reduce -every | view barchart`. I thought that since reduce/head/sort/etc would all support -every, they would explicitly set the batching interval (defaulting to "aggregate all" if there's no -every, i.e. effectively unbatching). Clearly that was not a complete understanding.

Considering mccanne's case of essentially not wanting to type extra -every in each downstream proc, and the newly revealed cases of having to `unbatch`, if we wished to support both without inconveniencing the user, I think we'd need a concept of batch being "persistent" (propagating downstream and affecting procs that come later). Since batch marks are now objects, perhaps that is doable. Then:

1. Use 'batch' to set the persistent batch interval for all downstream procs, which then don't specify -every (if they do, -every sets the new batching interval and unsets the persistent bit).

2. Use -every to affect only the current proc but not downstream ones; each downstream proc will specify -every if it's meant to be batching, or omit it for aggregate-all behavior. Note that reduce -every would still emit batch marks, so the sink could get them, but without persistent bit (that's how the downstream proc would know to override with its own -every or the 'all' default).

3. Use of 'unbatch' would still be supported so a program with upstream 'batch' could unset the batching behavior, but in case 2, one would not need to put 'unbatch' in the program.

Thoughts?

**Comment by dmehra** [comment]
@dmajda @demmer bump

I just realized that our new semantics killed my dev workflow of writing a program that ends in "head 3" so i see what kind of data is produced by my juttle, because something like `... | reduce -every :day: sum(value) | head 3` now shows me 3 points for each day in my program. As a result, I now have to include `unbatch` in 90% of juttles I type up. This needs to change.

**Comment by dmajda** [comment]
We discussed this with @demmer yesterday. The result is that we will revert the decision that `-every` emits marks.

The discussion we had with @mccanne a while ago was about whether batches should be “real”, that is, whether the stream *is a sequence of batches* or whether it is simply a sequence of points which are sometimes merely *viewed as batches*. Apparently, neither approach is ideal and sufficient by itself. If we always embed batches into the stream, we end up using `unbatch` all the time (the current situation). If we get rid of them, we end up repeating `-every` all the time and we will have problems with views.

Ultimately, we decided we need both approaches and the line between them should be `batch`/`unbatch` vs. `-every`. The former is “global” in that it embeds/removes batches to/from the stream, the latter is “local” in that it only influences the current proc.

It’s clear that not all workflows can be represented with these two concepts in an ideal way, but they provide solid building blocks with well-defined semantics. There is no hard-to-understand magic below (that would be the case with persistent/non-persistent batches as proposed by @dmehra).

I’ll morph this issue into an enhancement covering the reversal, @bkutil will do the actual work.
